sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1367,6 +1367,22 @@ class Dataset[T] private[sql](
     }: _*)
   }
 
+  /**
+   * Casts all the values of the current Dataset following the types of a specific StructType.
+   * This method works also with nested structTypes.
+   *
+   *  @group typedrel
+   *  @since 2.4.0
+   */
+  def castBySchema(schema: StructType): DataFrame = {
+    assert(schema.fields.map(_.name).toList.sameElements(this.schema.fields.map(_.name).toList),
+      "schema should have the same fields as the original schema")
+
+    selectExpr(schema.map(
```

Comment:
This is basically one line code. I don't think we need an API for the one liner.

Reply:
Hello @HyukjinKwon, Thanks for your feedback.
Actually, some methods are one liner in the current API, but they help users by allowing them to know what they can do via the API (printSchema, dtypes, columns are perfect examples for that).
Even if it's a one liner, this method can be helpful since it will tell spark's users that they can cast a dataframe by passing a schema.
This method can be very useful in the Big Data world; generally, we parse different files which are in strings, and we have to cast all these values following a schema. In my case, my client provides Avro schemas to define the columns and names.
I searched a lot for a method which applies a schema on a dataframe but I couldn't find one (https://stackoverflow.com/questions/51561715/cast-values-of-a-spark-dataframe-using-a-defined-structtype/51562763#51562763).
Even by searching on github, I always see examples where people cast each value separately.

Reply:
-1 (I think it is a pretty sensitive issue to add a new api in `Dataset`....)

Reply:
@maropu shall we add the @Experimental annotation in that case ? What do you suggest to make this method less sensitive ? Thanks

Reply:
most of one liner APIs should have been considered more before being added - this doesn't mean we are okay with one liner API.

You can just define it in application side and use it. Pretty easy and simple. You can answer some questions about casting. I don't think it's worth adding it. There are already too many APIs open.

Reply:
I am still not convinced about not adding this since it might be helpful for Spark users.
I totally agree about avoiding to add it to the Dataset Api, but can we add it to the functions object in the org.apache.spark.sql package for example ? 
Thanks

Reply:
There are many good one liner tricks and I would just leave those good tricks in mailing list or something. I wouldn't add an API only because it _might be_ helpful to some users.

We shouldn't add an API only because it _might be_ useful. I would consider adding this if there's a request for this PR multiple times, it is not one liner change and there's no easy workaround for it.

Otherwise, every system will have an API to send an email.

Reply:
Ok I understand, Thanks